File: app/src/main/java/com/example/krystianwsul/organizator/loaders/ShowInstanceLoader.java
package com.example.krystianwsul.organizator.loaders;

import android.content.Context;
import android.text.TextUtils;

import com.example.krystianwsul.organizator.domainmodel.DomainFactory;
import com.example.krystianwsul.organizator.utils.InstanceKey;

import junit.framework.Assert;

public class ShowInstanceLoader extends DomainLoader<ShowInstanceLoader.Data> {
    private final InstanceKey mInstanceKey;

    public ShowInstanceLoader(Context context, InstanceKey instanceKey) {
        super(context);

        Assert.assertTrue(instanceKey != null);

        mInstanceKey = instanceKey;
    }

    @Override
    public Data loadInBackground() {
        return DomainFactory.getDomainFactory(getContext()).getShowInstanceData(getContext(), mInstanceKey);
    }

    public static class Data extends DomainLoader.Data {
        public final InstanceKey InstanceKey;
        public final String Name;
        public final String DisplayText;
        public boolean Done;
        public final boolean TaskCurrent;
        public final boolean IsRootInstance;
        public final Boolean IsRootTask;

        public Data(InstanceKey instanceKey, String name, String displayText, boolean done, boolean taskCurrent, boolean isRootInstance, Boolean isRootTask) {
            Assert.assertTrue(instanceKey != null);
            Assert.assertTrue(!TextUtils.isEmpty(name));

            InstanceKey = instanceKey;
            Name = name;
            DisplayText = displayText;
            Done = done;
            TaskCurrent = taskCurrent;
            IsRootInstance = isRootInstance;
            IsRootTask = isRootTask;
        }

        @Override
        public int hashCode() {
            int hashCode = 0;
            hashCode += InstanceKey.hashCode();
            hashCode += Name.hashCode();
            if (!TextUtils.isEmpty(DisplayText))
                hashCode += DisplayText.hashCode();
            hashCode += (Done ? 1 : 0);
            hashCode += (TaskCurrent ? 1 : 0);
            hashCode += (IsRootInstance ? 1 : 0);
            if (IsRootTask != null)
            hashCode += (IsRootTask ? 2 : 1);
            return hashCode;
        }

        @SuppressWarnings("RedundantIfStatement")
        @Override
        public boolean equals(Object object) {
            if (object == null)
                return false;

            if (object == this)
                return true;

            if (!(object instanceof Data))
                return false;

            Data data = (Data) object;

            if (!InstanceKey.equals(data.InstanceKey))
                return false;

            if (!Name.equals(data.Name))
                return false;

            if (TextUtils.isEmpty(DisplayText) != TextUtils.isEmpty(data.DisplayText))
                return false;

            if (!TextUtils.isEmpty(data.DisplayText) && !DisplayText.equals(data.DisplayText))
                return false;

            if (Done != data.Done)
                return false;

            if (TaskCurrent != data.TaskCurrent)
                return false;

            if (IsRootInstance != data.IsRootInstance)
                return false;

            if ((IsRootTask == null) != (data.IsRootTask == null))
                return false;

            if (IsRootTask != null && !IsRootTask.equals(data.IsRootTask))
                return false;

            return true;
        }
    }
}

File: app/src/main/java/com/krystianwsul/checkme/loaders/ShowInstanceLoader.kt
package com.krystianwsul.checkme.loaders

import android.content.Context
import com.krystianwsul.checkme.domainmodel.DomainFactory
import com.krystianwsul.checkme.gui.instances.tree.GroupListFragment
import com.krystianwsul.checkme.utils.InstanceKey
import com.krystianwsul.checkme.utils.TaskKey


class ShowInstanceLoader(context: Context, private val instanceKey: InstanceKey) : DomainLoader<ShowInstanceLoader.DomainData>(context, if (instanceKey.type == TaskKey.Type.REMOTE) FirebaseLevel.NEED else FirebaseLevel.NOTHING) {

    override val name = "ShowInstanceLoader, instanceKey: $instanceKey"

    override fun loadDomain(domainFactory: DomainFactory) = domainFactory.getShowInstanceData(context, instanceKey)

    data class DomainData(val instanceData: InstanceData?) : com.krystianwsul.checkme.viewmodels.DomainData()

    data class InstanceData(val name: String, val displayText: String?, var done: Boolean, val taskCurrent: Boolean, val isRootInstance: Boolean, val exists: Boolean, val dataWrapper: GroupListFragment.DataWrapper) {

        init {
            check(name.isNotEmpty())
        }
    }
}
